packages/dds/tree/src/util/utils.ts

```diff
@@ -4,13 +4,34 @@
  */
 
 import structuredClone from "@ungap/structured-clone";
+import { Brand, Opaque } from "./brand";
 
 /**
  * Make all transitive properties in T readonly
  */
-export type RecursiveReadonly<T> = {
-    readonly [P in keyof T]: RecursiveReadonly<T[P]>;
-};
+export type RecursiveReadonly<T> = T extends
+    | undefined
+    // eslint-disable-next-line @rushstack/no-new-null
+    | null
+    | boolean
+    | string
+    | number
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    | Function
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    | Symbol
+    ? T
+    : T extends Opaque<Brand<any, string>>
+    ? T
```

Comment:
I initially had this:
```
    : T extends Opaque<Brand<infer V, string>>
    ? Opaque<Brand<RecursiveReadonly<V>, string>>
```
but that leads to the type of such branded members becoming `{}`.
Thoughts?

Reply:
don't look inside Opaque types. Their opaqueness is implemented by making their type not actually describe their data, so looking inside them is a bad idea. Opaque types are already readonly unless you do some explicit type conversion to unopaque them anyway since you can't mutate something with no real type information about it.

As for why it breaks in that specif way: I could figure it out if you really want, but I suspect its not worth it.

Reply:
Alright, sounds good. Resolving this thread then.